File: apps/web/src/lib/types/public-site.ts
// Public-facing site types — mirrors apps/api/schemas/public_site.py

export interface PublicSiteStats {
  listings_count: number;
  agents_count: number;
  featured_area: string | null;
  inventory_value_aed: number | null;
}

// ---------------------------------------------------------------------------
// Config sub-types — mirrors apps/api/schemas/public_site_config.py
// ---------------------------------------------------------------------------

export interface HeroConfig {
  headline: string | null;
  subheadline: string | null;
  eyebrow: string | null;
  image_url: string | null;
}

export interface ThemeConfig {
  primary_color: string | null;
  accent_color: string | null;
}

export interface ServiceCard {
  title: string;
  description: string;
}

export interface ServicesConfig {
  enabled: boolean;
  title: string | null;
  subtitle: string | null;
  cards: ServiceCard[] | null;
}

export interface TeamMember {
  name: string;
  title: string;
  photo_url: string | null;
  phone: string | null;
  email: string | null;
  linkedin_url: string | null;
}

export interface TeamConfig {
  title: string | null;
  subtitle: string | null;
  members: TeamMember[];
  show_agents: boolean;
}

export interface FooterConfig {
  description: string | null;
  address: string | null;
  instagram_url: string | null;
  facebook_url: string | null;
  linkedin_url: string | null;
}

export interface SectionsConfig {
  services_enabled: boolean;
  team_enabled: boolean;
  contact_enabled: boolean;
}

export interface ContactSectionConfig {
  title: string | null;
  headline: string | null;
  whatsapp: string | null;
}

export interface RawSiteConfig {
  hero?: Partial<HeroConfig> | null;
  theme?: Partial<ThemeConfig> | null;
  services?: Partial<ServicesConfig> | null;
  team?: Partial<TeamConfig> | null;
  footer?: Partial<FooterConfig> | null;
  sections?: Partial<SectionsConfig> | null;
  contact?: Partial<ContactSectionConfig> | null;
}

export interface ResolvedSiteConfig {
  hero: HeroConfig;
  theme: ThemeConfig;
  services: ServicesConfig;
  team: TeamConfig;
  footer: FooterConfig;
  sections: SectionsConfig;
  contact: ContactSectionConfig;
}

// ---------------------------------------------------------------------------
// Public profile
// ---------------------------------------------------------------------------

export interface PublicTenantProfile {
  name: string;
  logo_url: string | null;
  tagline: string | null;
  contact_email: string;
  contact_phone: string;
  /** ISO-3166-1 alpha-2 country codes where the tenant operates */
  operating_countries: string[];
  stats?: PublicSiteStats;
  config?: RawSiteConfig | null;
}

export interface PublicListingItem {
  id: string;
  title: string;
  short_description: string | null;
  price: number;
  currency: string;
  beds: number | null;
  baths: number | null;
  area_sqft: number | null;
  address: string | null;
  property_type: string;
  primary_photo_url: string | null;
  purpose: string;
  listing_tier?: string | null;
  created_at?: string | null;
}

export interface PublicListingListResponse {
  items: PublicListingItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface PublicAgentSnippet {
  id: string;
  full_name: string;
  photo_url: string | null;
  phone: string | null;
  email: string;
  license_id?: string | null;
}

export interface PublicListingDetail {
  id: string;
  title: string;
  short_description: string | null;
  description: string | null;
  price: number;
  currency: string;
  purpose: string;
  status: string;
  beds: number | null;
  baths: number | null;
  area_sqft: number | null;
  address: string | null;
  property_type: string;
  amenities: unknown[] | null;
  media_urls: string[];
  hero_media_url?: string | null;
  agent: PublicAgentSnippet | null;
  // Branding — always present; used by listing-detail page in direct_only mode
  // without requiring a separate profile fetch.
  tenant_name: string | null;
  tenant_logo_url: string | null;
  tenant_tagline: string | null;
  tenant_contact_email: string | null;
  tenant_contact_phone: string | null;
}

export interface PublicLeadCreate {
  name: string;
  email: string;
  phone?: string | null;
  message?: string | null;
  listing_id?: string | null;
}

export interface PublicLeadResponse {
  id: string;
}

// Admin: public-site settings
export interface PublicSiteSettings {
  slug: string;
  /** Platform master switch (read-only for tenants). */
  public_site_feature_enabled: boolean;
  public_site_enabled: boolean;
  /** When true, only listing-detail URLs resolve. */
  public_site_direct_links_only: boolean;
  public_site_logo_url: string | null;
  public_site_tagline: string | null;
  public_url_hint: string;
  config: RawSiteConfig;
  // Per-tenant media limits (read-only for tenant owners)
  max_images_per_property: number;
  max_videos_per_property: number;
  max_image_mb: number;
  max_video_mb: number;
}

export interface PublicSiteSettingsUpdate {
  public_site_enabled?: boolean | null;
  /** Tenant fine-grain: direct-link-only mode. */
  public_site_direct_links_only?: boolean | null;
  public_site_logo_url?: string | null;
  public_site_tagline?: string | null;
  config?: RawSiteConfig | null;
}

export interface PublicListingsParams {
  page?: number;
  page_size?: number;
  q?: string | null;
  purpose?: string | null;
  min_price?: number | null;
  max_price?: number | null;
  beds?: number | null;
  baths?: number | null;
  min_bedrooms?: number | null;
  max_bedrooms?: number | null;
  property_type?: string | null;
  city?: string | null;
  area?: string | null;
  amenities?: string | null;
  furnishing_status?: string | null;
  completion_status?: string | null;
  view_orientation?: string | null;
}

File: apps/api/schemas/public_site.py
"""Pydantic schemas for the public-site feature.

Covers:
- Public tenant profile (slug lookup)
- Public listing list + detail
- Public lead creation (inbound contact form)
- Tenant admin: read / update public-site settings
"""

import uuid
from typing import Any

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator

from apps.api.schemas.public_site_config import PublicSiteConfig


# ---------------------------------------------------------------------------
# Public: tenant stats (embedded in profile response)
# ---------------------------------------------------------------------------


class PublicTenantStats(BaseModel):
    listings_count: int
    agents_count: int
    featured_area: str | None = None
    inventory_value_aed: int


# ---------------------------------------------------------------------------
# Public: tenant profile
# ---------------------------------------------------------------------------


class PublicTenantProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name: str
    logo_url: str | None = None
    tagline: str | None = None
    contact_email: str
    contact_phone: str
    operating_countries: list[str] = Field(default_factory=lambda: ["AE"])
    stats: PublicTenantStats | None = None
    config: dict | None = None


# ---------------------------------------------------------------------------
# Public: listing list item (paginated)
# ---------------------------------------------------------------------------


class PublicOpenHouseSnippet(BaseModel):
    starts_at: str
    ends_at: str


class PublicListingItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    short_description: str | None = None
    price: float
    currency: str
    beds: int | None = None
    baths: int | None = None
    area_sqft: float | None = None
    address: str | None = None
    area: str | None = None
    city: str | None = None
    property_type: str
    primary_photo_url: str | None = None
    purpose: str
    tags: list[str] = Field(default_factory=list)
    is_verified: bool = False
    internal_reference: str | None = None
    price_per_sqft: float | None = None
    build_year: int | None = None
    listing_tier: str | None = None
    next_open_house_at: str | None = None
    # Agent card fields — populated by bulk query, no N+1.
    agent_name: str | None = None
    agent_phone: str | None = None
    agent_photo_url: str | None = None
    agent_has_license: bool = False


class PublicListingListResponse(BaseModel):
    items: list[PublicListingItem]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# Public: listing stats banner
# ---------------------------------------------------------------------------


class PublicListingTypeCount(BaseModel):
    property_type: str
    count: int


class PublicListingStatsResponse(BaseModel):
    total: int
    by_type: list[PublicListingTypeCount]


# ---------------------------------------------------------------------------
# Public: listing detail
# ---------------------------------------------------------------------------


class PublicAgentSnippet(BaseModel):
    id: uuid.UUID
    full_name: str
    photo_url: str | None = None
    phone: str | None = None
    email: str


class PublicUpcomingOpenHouse(BaseModel):
    starts_at: str
    ends_at: str
    capacity: int | None = None


class PublicListingDetailResponse(BaseModel):
    id: uuid.UUID
    title: str
    short_description: str | None = None
    description: str | None = None
    price: float
    currency: str
    purpose: str
    status: str
    beds: int | None = None
    baths: int | None = None
    area_sqft: float | None = None
    address: str | None = None
    area: str | None = None
    city: str | None = None
    internal_reference: str | None = None
    build_year: int | None = None
    property_type: str
    amenities: list[Any] | None = None
    tags: list[str] = Field(default_factory=list)
    media_urls: list[str] = Field(default_factory=list)
    floor_plan_urls: list[str] = Field(default_factory=list)
    hero_media_url: str | None = None
    agent: PublicAgentSnippet | None = None
    next_open_house: PublicOpenHouseSnippet | None = None
    upcoming_open_houses: list[PublicUpcomingOpenHouse] = Field(default_factory=list)
    # Off-plan delivery info (migration 0039).
    handover_year: int | None = None
    handover_quarter: int | None = None
    payment_plan: bool = False
    # Branding — present in all visibility states (including direct_only)
    # so the listing-detail page can render without a separate profile call.
    tenant_name: str | None = None
    tenant_logo_url: str | None = None
    tenant_tagline: str | None = None
    tenant_contact_email: str | None = None
    tenant_contact_phone: str | None = None


# ---------------------------------------------------------------------------
# Public: agents grid
# ---------------------------------------------------------------------------


class PublicAgentCard(BaseModel):
    id: uuid.UUID
    full_name: str
    email: str
    phone: str | None = None
    photo_url: str | None = None
    license_id: str | None = None


class PublicAgentsResponse(BaseModel):
    items: list[PublicAgentCard]


# ---------------------------------------------------------------------------
# Public: lead creation
# ---------------------------------------------------------------------------


class PublicLeadCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    email: EmailStr
    phone: str | None = Field(default=None, max_length=32)
    message: str | None = Field(default=None, max_length=2000)
    listing_id: uuid.UUID | None = None


class PublicLeadResponse(BaseModel):
    id: uuid.UUID


# ---------------------------------------------------------------------------
# Tenant admin: public-site settings
# ---------------------------------------------------------------------------


class PublicSiteSettingsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    slug: str
    public_site_feature_enabled: bool = False
    public_site_enabled: bool
    public_site_direct_links_only: bool = False
    public_site_logo_url: str | None = None
    public_site_tagline: str | None = None
    public_url_hint: str
    config: dict = Field(default_factory=dict)
    # Per-tenant media limits (read-only for tenant owners)
    max_images_per_property: int = 20
    max_videos_per_property: int = 2
    max_image_mb: int = 10
    max_video_mb: int = 25


class PublicSiteSettingsUpdate(BaseModel):
    """PATCH /tenants/me/public-site — all fields optional.

    Tenants may set public_site_enabled and public_site_direct_links_only.
    public_site_feature_enabled is platform-only; sending it raises 422.
    """

    public_site_enabled: bool | None = None
    public_site_direct_links_only: bool | None = None
    public_site_logo_url: str | None = None
    public_site_tagline: str | None = Field(default=None, max_length=200)
    config: PublicSiteConfig | None = None

    @field_validator("public_site_logo_url", mode="before")
    @classmethod
    def _validate_logo_url(cls, v: str | None) -> str | None:
        if v is None or v == "":
            return None
        if not v.startswith("https://"):
            raise ValueError("logo_url must use https://")
        return v
